lib/compat/wordpress-6.5/navigation-block-hooks.php

```diff
@@ -0,0 +1,119 @@
+<?php
+add_filter( 'rest_prepare_wp_navigation', 'gutenberg_hook_blocks_into_core_navigation_rest_response', 10, 3 );
+add_filter( 'block_core_navigation_render_inner_blocks', 'gutenberg_hook_blocks_into_core_navigation', 10, 3 );
+add_action( 'rest_insert_wp_navigation', 'gutenberg_update_ignore_hooked_blocks_meta_core_navigation', 10, 3 );
+
+/**
+ * Updates the post meta with the list of ignored hooked blocks when the navigation is created or updated via the REST API.
+ *
+ * @param WP_Post $post Post object.
+ */
+function gutenberg_update_ignore_hooked_blocks_meta_core_navigation( $post ) {
+	if ( ! isset( $post->ID ) ) {
+		return;
+	}
+
+	$anchor_block_type    = 'core/navigation';
+	$first_child_position = 'first_child';
+	$last_child_position  = 'last_child';
+
+	$hooked_blocks                 = get_hooked_blocks();
+	$hooked_block_types_last_child = isset( $hooked_blocks[ $anchor_block_type ][ $last_child_position ] )
+	? $hooked_blocks[ $anchor_block_type ][ $last_child_position ]
+	: array();
+
+	$hooked_block_types_first_child = isset( $hooked_blocks[ $anchor_block_type ][ $first_child_position ] )
+	? $hooked_blocks[ $anchor_block_type ][ $first_child_position ]
+	: array();
+
+	$hooked_block_types_last_child  = apply_filters( 'hooked_block_types', $hooked_block_types_last_child, $last_child_position, $anchor_block_type, null );
+	$hooked_block_types_first_child = apply_filters( 'hooked_block_types', $hooked_block_types_first_child, $first_child_position, $anchor_block_type, null );
```

Comment:
Nice! 🎉 

Alternatively, we could consider invoking the "full" Blocks Hooks logic (i.e. `make_before_block_visitor`, `make_after_block_visitor`, `traverse_and_serialize_blocks`) which would add the `ignoredHookedBlocks` attribute to the mock Navigation block, from which we could then extract it.

The major difference would be that we'd invoke Block Hooks at a higher level, including any changes made later to that logic (e.g. the new [`hooked_block_{$block_type}` filter](https://github.com/WordPress/wordpress-develop/pull/5835) that I'm about to introduce in WP 6.5), which they won't if we're "only" running lower-level methods.

OTOH, this might be a bit academic, since I'm also planning to port the functionality to inject hooked blocks into the Navigation block to WP 6.5.

So I think we can leave this as-is 😊 

...except for one minor request: I'd still love if we could pass the Navigation `$post` object as context here. Filters might want to access that 😬 

```suggestion
	$hooked_block_types_last_child  = apply_filters( 'hooked_block_types', $hooked_block_types_last_child, $last_child_position, $anchor_block_type, $post );
	$hooked_block_types_first_child = apply_filters( 'hooked_block_types', $hooked_block_types_first_child, $first_child_position, $anchor_block_type, $post );
```

Reply:
Updated in https://github.com/WordPress/gutenberg/pull/57754/commits/30dcb6470410687e25220113c13e1eb0d187435d